### PR TITLE
JAMES-2586 postgres users dao and repository

### PR DESCRIPTION
**Pull request by hungphan227**

### CODE REVIEW

File: server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUserModule.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        Field<String> HASHED_PASSWORD = DSL.field("hashed_password", SQLDataType.VARCHAR.notNull());
+        Field<String> ALGORITHM = DSL.field("algorithm", SQLDataType.VARCHAR(100).notNull());
+
+        PostgresTable TABLE = PostgresTable.name(TABLE_NAME.getName())
+            .createTableStep(((dsl, tableName) -> dsl.createTable(tableName)
+                .column(USERNAME)
+                .column(HASHED_PASSWORD)
+                .column(ALGORITHM)
+                .constraint(DSL.primaryKey(USERNAME))))
+            .supportsRowLevelSecurity();
```
Comment (hungphan227) [inline server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUserModule.java:44]: @Arsnael should we apply RLS to the user table?

File: server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUserModule.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+public interface PostgresUserModule {
+    interface PostgresUserTable {
+        Table<Record> TABLE_NAME = DSL.table("users");
+
+        Field<String> USERNAME = DSL.field("username", SQLDataType.VARCHAR(255).notNull());
+        Field<String> HASHED_PASSWORD = DSL.field("hashed_password", SQLDataType.VARCHAR.notNull());
+        Field<String> ALGORITHM = DSL.field("algorithm", SQLDataType.VARCHAR(100).notNull());
+
+        PostgresTable TABLE = PostgresTable.name(TABLE_NAME.getName())
+            .createTableStep(((dsl, tableName) -> dsl.createTable(tableName)
```
Comment (vttranlina) [inline server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUserModule.java:39]:
```suggestion
            .createTableStep(((dsl, tableName) -> dsl.createTableIfNotExists(tableName)
```
more strong

File: server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersRepository.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.user.postgres;
+
+import org.apache.james.domainlist.api.DomainList;
+import org.apache.james.user.lib.UsersRepositoryImpl;
+
+public class PostgresUsersRepository extends UsersRepositoryImpl<PostgresUsersDAO> {
+    public PostgresUsersRepository(DomainList domainList, PostgresUsersDAO usersDAO) {
```
Comment (Arsnael) [inline server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersRepository.java:26]: I would think you probably need an `@Inject` here :)

File: server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersDAO.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    @Override
+    public void updateUser(User user) throws UsersRepositoryException {
+        Preconditions.checkArgument(user instanceof DefaultUser);
+        DefaultUser defaultUser = (DefaultUser) user;
+
+        boolean executed = postgresExecutor.executeRow(dslContext -> Mono.from(dslContext.update(TABLE_NAME)
+                .set(HASHED_PASSWORD, defaultUser.getHashedPassword())
+                .set(ALGORITHM, defaultUser.getHashAlgorithm().asString())
+                .where(USERNAME.eq(user.getUserName().asString()))
+                .returning(USERNAME)))
```
Comment (chibenwa) [inline server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersDAO.java:77]: Why do we need `.returning(USERNAME)` ?

File: server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersDAO.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+            .map(record -> Username.of(record.get(USERNAME)));
+    }
+
+    @Override
+    public void addUser(Username username, String password) {
+        DefaultUser user = new DefaultUser(username, algorithm, algorithm);
+        user.setPassword(password);
+
+        postgresExecutor.executeVoid(dslContext -> Mono.from(dslContext.insertInto(TABLE_NAME, USERNAME, HASHED_PASSWORD, ALGORITHM)
+                .values(user.getUserName().asString(), user.getHashedPassword(), user.getHashAlgorithm().asString())))
```
Comment (chibenwa) [inline server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersDAO.java:132]: Indent please

File: server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersDAO.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+            .onErrorMap(e -> e instanceof DataAccessException && e.getMessage().contains(DUPLICATE_VIOLATION_MESSAGE),
```
Comment (chibenwa) [inline server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersDAO.java:133]:
```suggestion
e -> e instanceof DataAccessException && e.getMessage().contains(DUPLICATE_VIOLATION_MESSAGE)
```

Could we have a PG utility wrapping this common tricky condition?

File: server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersDAO.java (hunk 5 above)
Comment (hungphan227) [reply server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersDAO.java:132]: ? I do not see anything wrong

Comment (chibenwa) [reply server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersDAO.java:132]: This is IMO close to the limit for line length, but let say I'd let it pass this time.

File: server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersDAO.java (hunk 6 above)
Comment (Arsnael) [reply server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersDAO.java:133]: Or maybe can we consider it's enough on DataAccessException without having to check the content of the message?

Comment (vttranlina) [reply server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersDAO.java:133]: But we are not sure whenever get `DataAccessException`, it is exactly a "duplicate data" error.

Comment (chibenwa) [reply server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersDAO.java:133]: IMO I would be glad to read `onErrorResume(PGPredicate.ALREADY_EXIST_EXCEPTION, e -> new AlreadyExistInUsersRepositoryException("User with username " + username + " already exist!"))` and not get that duplicated everywhere.

Comment (Arsnael) [reply server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersDAO.java:133]: Hmmm... get your point. So stupidly verbose though :)

Comment (hungphan227) [reply server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersDAO.java]: fixed

File: server/apps/postgres-app/src/main/java/org/apache/james/PostgresJamesServerMain.java [hunk 7]
```diff
@@ -85,6 +86,7 @@ public class PostgresJamesServerMain implements JamesServerMain {
         new JPADataModule(),
         new JPAMailboxModule(),
         new PostgresMailboxModule(),
+        new PostgresDataModule(),
```
[the remark this answers is not in the document]
Comment (hungphan227) [reply server/apps/postgres-app/src/main/java/org/apache/james/PostgresJamesServerMain.java:87]: removed

File: server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresRepositoryConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+package org.apache.james.user.postgres;
+
+import org.apache.commons.configuration2.HierarchicalConfiguration;
+import org.apache.commons.configuration2.ex.ConfigurationException;
+import org.apache.commons.configuration2.tree.ImmutableNode;
+import org.apache.james.user.lib.model.Algorithm;
+import org.apache.james.user.lib.model.Algorithm.HashingMode;
+
+public class PostgresRepositoryConfiguration {
```
Comment (quantranhong1999) [inline server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresRepositoryConfiguration.java:28]: PostgresUsersRepositoryConfiguration would be a better name IMO.

File: server/container/guice/postgres-common/src/main/java/org/apache/james/modules/data/PostgresDataModule.java [hunk 9]
```diff
@@ -16,29 +16,14 @@
...
-import org.apache.james.utils.InitializationOperation;
-import org.apache.james.utils.InitilizationOperationBuilder;
+package org.apache.james.modules.data;
 
 import com.google.inject.AbstractModule;
-import com.google.inject.Scopes;
-import com.google.inject.multibindings.ProvidesIntoSet;
 
-public class JPAUsersRepositoryModule extends AbstractModule {
+public class PostgresDataModule extends AbstractModule {
```
Comment (quantranhong1999) [inline server/container/guice/postgres-common/src/main/java/org/apache/james/modules/data/PostgresDataModule.java:24]: `JPAUsersRepositoryModule` -> rename to `PostgresUsersRepositoryModule`

`JPADataModule` ->  rename to `PostgresDataModule`

would be more clear.

File: server/apps/postgres-app/src/main/java/org/apache/james/PostgresJamesServerMain.java [hunk 10]
```diff
@@ -114,8 +114,6 @@ public static void main(String[] args) throws Exception {
 
     static GuiceJamesServer createServer(PostgresJamesConfiguration configuration) {
         return GuiceJamesServer.forConfiguration(configuration)
-            .combineWith(POSTGRES_MODULE_AGGREGATE)
-            .combineWith(new UsersRepositoryModuleChooser(new JPAUsersRepositoryModule())
-                .chooseModules(configuration.getUsersRepositoryImplementation()));
+            .combineWith(POSTGRES_MODULE_AGGREGATE);
```
Comment (quantranhong1999) [inline server/apps/postgres-app/src/main/java/org/apache/james/PostgresJamesServerMain.java:117]: We need to keep the user module chooser IMO.

```suggestion
            .combineWith(new UsersRepositoryModuleChooser(new PostgresUsersRepositoryModule())
                .chooseModules(configuration.getUsersRepositoryImplementation()));
```

This should solve the failing test `userFromLdapShouldLoginViaImapProtocol(GuiceJamesServer) – org.apache.james.PostgresWithLDAPJamesServerTest`

**Comment by hungphan227** [comment]
https://ci-builds.apache.org/blue/organizations/jenkins/james%2FApacheJames/detail/PR-1803/7/tests

The test that check the concurrency issue of postgres connection FAILED

**Comment by Arsnael** [comment]
https://github.com/Arsnael/james-project/blob/domain-postgres/server/apps/postgres-app/src/main/resources/META-INF/persistence.xml#L35

Can we remove this line as well? Otherwise looks good to me :)

**Review by vttranlina — APPROVED** [review]

**Review by quantranhong1999 — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Pull request merged**